qualcommax: ipq60xx: add support for Cambium Networks XE3-4

skramstad opened the pull request:
Cambium Networks XE3-4 is a tri-radio Wi-Fi 6/6E 4×4/2×2 AP.
```
Hardware:
    Model:    Cambium Networks XE3-4
    CPU:      IPQ6010/AP-CP01-C3, SoC Version: 1.0 @ 800 MHz
    Memory:   1 GiB
    Flash:    512 MiB Macronix MX30UF2G18AC + W25Q128FW
    Ethernet: 1x 1 GbE   (QCA8072)
              1x 2.5 GbE (QCA8081)
    Buttons:  1x Reset
    Serial:   TX, RX, GND
    Baudrate: 115200
    Radios:   Qualcomm Atheros IPQ6018 802.11ax - 2x2 - 2GHz
              Qualcomm Atheros IPQ6018 802.11ax - 2x2 - 5GHz
              Qualcomm Atheros QCN9074 802.11ax - 4x4 - 5GHz or 6GHz
              BLE 4.1
    Power:    32.0W 802.3bt5 PoE++
              25.5W 802.3at with USB, BT disabled
    Size:     215mm x 215mm
    Ports:    1x USB 2.0
    Antenna:  6 GHz: 6.29 dBi, Omni    30 dBm
              5 GHz: 6.12 dBi, Omni    31 dBm
              2.4 GHz: 4.85 dBi, Omni  29 dBm
    LEDs:     Multi-color status LEDs
    Mounting: Wall, ceiling or T-bar
```
**Installation: Serial connection**
```
1. Open the AP to get access to the board. Connect RX, TX and GND.
2. Power on the AP, and short the CS pin of the SPI flash with one of the APs GND pins.
3. Transfer the initramfs image with TFTP 
   (Default server IP is 192.168.0.120) 
   # tftpboot factory.ubi
5. Flash the rootfs partition 
   # flash rootfs
7. Reboot the AP 
   # reset
```

Signed-off-by: Kristian Skramstad <kristian+github@83.no>

skramstad committed:
qualcommax: ipq60xx: add support for Cambium Networks XE3-4

Cambium Networks XE3-4 is a tri-radio Wi-Fi 6/6E 4×4/2×2 AP.

Hardware:
    Model:    Cambium Networks XE3-4
    CPU:      IPQ6010/AP-CP01-C3, SoC Version: 1.0 @ 800 MHz
    Memory:   1 GiB
    Flash:    512 MiB Macronix MX30UF2G18AC + W25Q128FW
    Ethernet: 1x 1 GbE   (QCA8072)
              1x 2.5 GbE (QCA8081)
    Buttons:  1x Reset
    Serial:   TX, RX, GND
    Baudrate: 115200
    Radios:   Qualcomm Atheros IPQ6018 802.11ax - 2x2 - 2GHz
              Qualcomm Atheros IPQ6018 802.11ax - 2x2 - 5GHz
              Qualcomm Atheros QCN9074 802.11ax - 4x4 - 5GHz or 6GHz
              BLE 4.1
    Power:    32.0W 802.3bt5 PoE++
              25.5W 802.3at with USB, BT disabled
    Size:     215mm x 215mm
    Ports:    1x USB 2.0
    Antenna:  6 GHz: 6.29 dBi, Omni    30 dBm
              5 GHz: 6.12 dBi, Omni    31 dBm
              2.4 GHz: 4.85 dBi, Omni  29 dBm
    LEDs:     Multi-color status LEDs
    Mounting: Wall, ceiling or T-bar

Installation: Serial connection
1. Open the AP to get access to the board. Connect RX, TX and GND.
2. Power on the AP, and short the CS pin of the SPI flash with
   one of the APs GND pins.
3. Transfer the initramfs image with TFTP
   (Default server IP is 192.168.0.120)
   # tftpboot factory.ubi
4. Flash the rootfs partition
   # flash rootfs
5. Reboot the AP
   # reset

Signed-off-by: Kristian Skramstad <kristian+github@83.no>
Link: https://github.com/openwrt/openwrt/pull/15633
Signed-off-by: Robert Marko <robimarko@gmail.com>

skramstad commented:
Note: BDF pull here https://github.com/openwrt/firmware_qca-wireless/pull/43 but trying to figure out why the qcn9074 BDF is not working for 5GHz or 6GHz.

skramstad commented:
@robimarko I think this is ready to be reviewed, and I've added the working 5GHz BDF for QCN9074.

One thing I'm not sure about, if it's OK to do this:
(I have to add this or `env` did not work).
```
&qpic_nand {

	nand@0 {

		partitions {
```
since it's already added in `ipq6018-cp01-c1.dts`
```
&qpic_nand {
	status = "okay";

	nand@0 {
		reg = <0>;

		nand-ecc-strength = <4>;
		nand-ecc-step-size = <512>;
		nand-bus-width = <8>;
	};
};
```

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-xe3-4.dts:
```diff
@@ -0,0 +1,402 @@
+};
+
+&qpic_nand {
+
```
Remove this newline

skramstad replied:
Done.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-xe3-4.dts:
```diff
@@ -0,0 +1,402 @@
+&qpic_nand {
+
+	nand@0 {
+
```
Remove this newline

skramstad replied:
Done.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-xe3-4.dts:
```diff
@@ -0,0 +1,402 @@
+};
+
+&blsp1_spi1 {
+
```
Rmove this newline

skramstad replied:
Done.

skramstad commented:
@robimarko should this PR be merged so future devices can also use the `ipq6018-fixed-smps.dtsi` if needed?

If something need to be fixed let me know.

robimarko commented on package/boot/uboot-envtools/files/qualcommax_ipq60xx:
```diff
@@ -9,7 +9,8 @@ board=$(board_name)
 8devices,mango-dvk|\
-8devices,mango-dvk-sfp)
+8devices,mango-dvk-sfp|\
+cambiumnetworks,xe3-4)
```
You dont really need this because you added the U-Boot compatible to the environment partition, uboot-envtools will find that automatically.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-xe3-4.dts:
```diff
@@ -0,0 +1,399 @@
+	label = "lan1";
+};
+
+&blsp1_spi1 {
```
Missing pinctrl and cs-select

skramstad replied:
I'm importing `#include "ipq6018-cp01-c1.dts"` https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/qcom/ipq6018-cp01-c1.dts , but that does not count?

When I setup the DTS I saw that most of what I needed was in that dts file.

robimarko replied:
Oh, that would explain it, but please dont include that and instead make a standalone DTS.
CP01 is a reference board and while yours is based on it they are not related and I dont want to have to untangle upstream DTS changes when something breaks later.

skramstad replied:
Roger that, I didn't know that. I'll fix that. That goes for the USB too I recon.

robimarko replied:
Yeah, completely remove the include on that DTS and add anything that you relied on to your DTS directly.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-xe3-4.dts:
```diff
@@ -0,0 +1,399 @@
+};
+
+&blsp1_spi1 {
+	flash@0 {
```
Missing reg and compatible at least, you want to set the `spi-max-frequency` as well

robimarko commented on target/linux/qualcommax/image/ipq60xx.mk:
```diff
@@ -11,6 +11,20 @@ define Device/8devices_mango-dvk
+       PAGESIZE := 2048
+       DEVICE_DTS_CONFIG := config@cp01-c3-xv3-4
+       SOC := ipq6010
+       DEVICE_PACKAGES := -kmod-usb3 -kmod-usb-dwc3 -kmod-usb-dwc3-qcom kmod-usb2 \
```
There isn't really a reason for dropping the USB kmods.

BTW, is the USB2.0 port even enabled as I dont see it being described?

skramstad replied:
It's enabled from the imported file as I mentioned earlier.

```
#include "ipq6018-cp01-c1.dts"
...
&qusb_phy_1 {
	status = "okay";
};

&usb2 {
	status = "okay";
};
```

robimarko commented on target/linux/qualcommax/ipq60xx/base-files/etc/hotplug.d/firmware/11-ath11-caldata:
```diff
@@ -20,6 +23,14 @@ case "$FIRMWARE" in
 	esac
 	;;
+"ath11k/QCN9074/hw1.0/cal-pci-0000:01:00.0.bin"|\
+"ath11k/QCN9074/hw1.0/cal-pci-0001:01:00.0.bin")
```
Considering there is only one PCIe slot, only one of these two (Probably the first one) is valid

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6010-xe3-4.dts:
```diff
@@ -0,0 +1,399 @@
+				reg = <0x510000 0x10000>;
+
+				ethaddr: ethaddr {
+					#nvmem-cell-cells = <1>;
```
This should probably be `#nvmem-cell-cells = <0>;` since you are not incrementing them in the only uses.

skramstad commented:
@robimarko, pushed latest changes. Local build was OK.

skramstad commented:
@robimarko after I removed env from `package/boot/uboot-envtools/files/qualcommax_ipq60xx`
```
Can't open /sys/bus/nvmem/devices/u-boot-env0/nvmem: Permission denied
Error: can't write fw_env to flash
```

robimarko commented:
@skramstad Nope, its my fault, it makes envtools use the underlying NVMEM device and the current driver does not have write support, only read support so that is why it fails, if you need to be able to write then we still need to use the old school configuration via script.

skramstad commented:
@robimarko Ok, so adding back this code is the old school way?
```
idx="$(find_mtd_index 0:APPSBLENV)"
[ -n "$idx" ] && \
	ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x10000" "0x10000"
```

robimarko commented:
@skramstad Yes

skramstad commented:
Thanks. Already pushed new commit.

skramstad commented:
@robimarko I'm still having problem writing to env after adding back the old school way.

robimarko commented:
Did it work before?

If I am reading the `envtools` code correctly, it will first try to use the `/etc/fw_env.config` and only then the NVMEM device.
So, make sure that its present and populated.

skramstad commented:
@robimarko Yes, it did work before. Before I was using the commit from when I created the PR. I believe the last working one was commit `398f06499d3b9e1731e7d8aa2d22313078951c10` from June 1.

I'll see what else that could cause this, and also as you said, make sure the env files are present and populated. I can read the env with `fw_printenv` but not write.

skramstad commented:
@robimarko Hi again. Got it working with latest PR. Why it suddenly didn't work when I added back the "old school way" was failed sysupgrade. I think that was related to not able to set fw_setenv. I also added the missing uart in DTS.

So I think this PR is done 👍

robimarko commented:
Thanks! Rebased on top of main and merged!
